source/Octopus.Tentacle/Kubernetes/Diagnostics/KubernetesAgentMetrics.cs

```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Util;
+
+namespace Octopus.Tentacle.Kubernetes.Diagnostics
+{
+    public class KubernetesAgentMetrics
+    {
+        readonly string lastEventTimestampKey = "latestTimestamp";
+
+        public delegate KubernetesAgentMetrics Factory(IPersistenceProvider persistenceProvider);
+
+        readonly IPersistenceProvider persistenceProvider;
+        readonly ISystemLog log;
+
+        public KubernetesAgentMetrics(IPersistenceProvider persistenceProvider, ISystemLog log)
+        {
+            this.persistenceProvider = persistenceProvider;
+            this.log = log;
+        }
+
+        public void TrackEvent(string reason, string source, DateTimeOffset occurrence)
+        {
+            try
+            {
+                lock (persistenceProvider)
+                {
+                    var sourceEventsForReason = LoadFromPersistence(reason) ?? new Dictionary<string, List<DateTimeOffset>>();
+
+                    if (!sourceEventsForReason.TryGetValue(source, out var occurenceTimestamps))
+                    {
+                        occurenceTimestamps = new List<DateTimeOffset>();
+                        sourceEventsForReason[source] = occurenceTimestamps;
+                    }
+
+                    occurenceTimestamps.Add(occurrence);
+                    Persist(reason, sourceEventsForReason);
+                    UpdateLatestTimestamp(occurrence);
+                }
+            }
+            catch (Exception e)
+            {
+                log.Error($"Failed to persist a kubernetes event metric, {e.Message}");
+            }
+        }
+
+        public DateTimeOffset GetLatestEventTimestamp()
+        {
+            lock (persistenceProvider)
+            {
+                return GetLatestEventTimeStampInternal();
+            }
+        }
+
+        DateTimeOffset GetLatestEventTimeStampInternal()
+        {
+            //NOTE: this must be called from within a lock on the PersistenceProvider.
+            var timeStampString = persistenceProvider.GetValue(lastEventTimestampKey);
+            if (!timeStampString.IsNullOrEmpty())
+            {
+                return DateTimeOffset.Parse(timeStampString!);    
+            }
+            return DateTimeOffset.MinValue;
+        }
+
+        void UpdateLatestTimestamp(DateTimeOffset newEventTime)
+        {
+            try
+            {
+                var latestEvent = GetLatestEventTimestamp();
```

Comment:
```suggestion
                var latestEvent = GetLatestEventTimestampInternal();
```

Reply:
I saw the comment from Kevin but I would probably just call the public method and not have the internal one at all. Much of a muchness though 🤷

Reply:
likewise - it all works :shrug: